compiler/rustc_attr/src/session_diagnostics.rs

```diff
@@ -223,8 +225,10 @@ impl<'a> SessionDiagnostic<'a> for UnsupportedLiteral {
             error_code!(E0565),
         );
         if self.is_bytestr {
+            let start_point = handler.span_start_point_from_emitter(self.span).unwrap_or(self.span);
```

Comment:
Maybe we could expose like `fn opt_source_map(&self) -> Option<&SourceMap>` or something on `Handler`? Not sure that's the right direction, but just a thought so we don't need to expose such specific functions like `start_point` from the handler itself.

Reply:
Oh -- it's because of that `borrow` call in `inner.borrow().emitter.source_map()`, I guess the lifetimes may not work out? :thinking:

Maybe something like `fn with_opt_source_map<T>(&self, impl Fn(&SourceMap) -> T) -> Option<T>` Then we do:

```
let start_point = handler.with_opt_source_map(|sm| sm.start_point(self.span)).flatten().unwrap_or(self.span);
```

Reply:
I agree with this, I'd prefer not to have `SourceMap`'s methods on `Handler`. `opt_source_map` or `with_opt_source_map` would be okay, or we could add a `&'a SourceMap` field (using `#[skip_arg]`) in the specific diagnostic struct that uses those functions.

Reply:
Addressed your comments here https://github.com/rust-lang/rust/pull/101434/commits/1524b59444109cf71916653d7906e4d41bd4ba6e.
Chose the path of having a `SourceMap` in the struct because it guarantees we keep the existing flow, and don't have to fallback to `self.span`.